src/coord/src/coord.rs

```diff
@@ -1259,6 +1281,52 @@ impl<S: Append + 'static> Coordinator<S> {
             }) => {
                 // TODO(guswynn): communicate `bindings` to `sequence_peek`
             }
+            ControllerResponse::ComputeReplicaHeartbeat(replica_id, when) => {
+                // Round off the time based on `REPLICA_STATUS_GRANULARITY`.
+                // We can't just do something like `(timestamp / REPLICA_STATUS_GRANULARITY) * REPLICA_STATUS_GRANULARITY`
+                // with a Unix timestamp, because then everything will be offset due to leap seconds
+                // (e.g. rounding to 60 seconds will produce records at every HH:mm:26 since
+                //  there have been 26 leap seconds since 1970).
+                // This wouldn't be a real problem but would look a bit annoying.
+                let when_coarsened = {
+                    const REPLICA_STATUS_GRANULARITY: u32 = 60; /* seconds, must be less than 1 day */
+                    let seconds = when.time().num_seconds_from_midnight();
+                    let seconds_offset = seconds % REPLICA_STATUS_GRANULARITY;
+                    (when - chrono::Duration::seconds(seconds_offset.into()))
+                        .with_nanosecond(0)
+                        .unwrap()
+                };
+                let new = ReplicaMetadata {
+                    last_heartbeat: when_coarsened,
+                };
+                let old = self
+                    .transient_replica_metadata
+                    .insert(replica_id, new.clone());
+
+                if old.as_ref() != Some(&new) {
+                    let table = self
+                        .catalog
+                        .state()
+                        .resolve_builtin_table(&MZ_CLUSTER_REPLICA_HEARTBEATS);
+                    let retraction = old.map(|old| BuiltinTableUpdate {
+                        id: table,
+                        row: old.as_row(replica_id),
+                        diff: -1,
+                    });
+                    let insertion = BuiltinTableUpdate {
+                        id: table,
+                        row: new.as_row(replica_id),
+                        diff: 1,
+                    };
+
+                    let updates = if let Some(retraction) = retraction {
+                        vec![retraction, insertion]
+                    } else {
+                        vec![insertion]
+                    };
```

Comment:
I think this logic, and some of the other logic that generates `BuiltinTableUpdate`s belongs in `builtin_table_updates.rs`.